### PR TITLE
Add random name button to Ship Info panel's rename dialog

### PR DESCRIPTION
**Pull request by nickshanks**
# Feature

## Summary
When renaming a ship using the Ship Info panel (launch game, press 'E I S Shift-R'), provide a button to suggest a random name. I needed this because I had lots of ships with names assigned that were no longer correct (e.g. "Heavy Laser Bastion VIII" now had Sunbeams on it) and I couldn't be arsed renaming them all manually, especially as the arrow keys don't work in the text input field (another bug!).

I also moved it to the left edge, and renamed it, because for _ages_ I had assumed clicking it would close the dialog and assign a random name to the ship, which I didn't want to do, rather than merely suggest a name.

## Screenshots
Before:
Screenshot 2024-06-20 at 13 01 34
After:
Screenshot 2024-06-20 at 13 00 49

## Testing Done
Renamed a ship I already had, bought a ship with a random name, bought 500 shuttles and left the name field blank.

## Save File
Any save will do.

## Wiki Update
Doubt it, perhaps button rename might affect this?

## Performance Impact
N/A

### CODE REVIEW

File: source/NameDialog.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/* ShipyardPanel.cpp
```
Comment (eebop) [inline source/NameDialog.cpp:1]:
```suggestion
/* NameDialog.cpp
```

File: source/NameDialog.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/* NameDialog.cpp
+Copyright (c) 2024 Endless Sky contributors
```
Comment (eebop) [inline source/NameDialog.cpp:2]:
```suggestion
Copyright (c) 2024 by Endless Sky contributors
```

File: source/NameDialog.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
+/* NameDialog.h
+Copyright (c) 2024 Endless Sky contributors
```
Comment (eebop) [inline source/NameDialog.h:2]:
```suggestion
Copyright (c) 2024 by Endless Sky contributors
```

**Review by eebop — COMMENTED** [review]
The check is really finicky

**Comment by nickshanks** [comment]
> The check is really finicky 

I would rather the check was changed. The word "by" is not permitted in legally valid copyright statements.

**Comment by TheGiraffe3** [comment]
Well, you could use nickshanks, or if you have any other usernames that would work. You’re right about `by` though. 
Maybe shannick? Or Endless Sky Contributor?

File: source/NameDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+	randomPos = cancelPos - Point(100., 0.);
+	SpriteShader::Draw(SpriteSet::Get("ui/dialog cancel"), randomPos);
+
+	const Font &font = FontSet::Get(14);
+	static const string RANDOM = "Suggest";
+	Point labelPos = randomPos - .5 * Point(font.Width(RANDOM), font.Height());
+	font.Draw(RANDOM, labelPos, *GameData::Colors().Get("medium"));
+}
+
```
Comment (TomGoodIdea) [inline source/NameDialog.cpp:36]:
```suggestion



```

File: source/NameDialog.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "Dialog.h"
+
+#include "GameData.h"
+#include "Phrase.h"
+#include "Point.h"
+
+#include <string>
+
+using namespace std;
```
Comment (TomGoodIdea) [inline source/NameDialog.h:27]: we don't do this in headers
```suggestion

```

File: source/NameDialog.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+*/
+
+#include "NameDialog.h"
+
+#include "Color.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+
```
Comment (TomGoodIdea) [inline source/NameDialog.cpp:23]:
```suggestion
using namespace std;



```

File: source/NameDialog.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+// The name entry dialog should include a "Random" button to choose a random
+// name using the civilian ship name generator.
+class NameDialog : public Dialog {
+public:
+	template <class T>
+	NameDialog(T *panel, void (T::*fun)(const string &), const string &message, string initialValue = "")
```
Comment (TomGoodIdea) [inline source/NameDialog.h:34]:
```suggestion
	NameDialog(T *panel, void (T::*fun)(const string &), const std::string &message, std::string initialValue = "")
```

File: source/NameDialog.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+		: Dialog(panel, fun, message, initialValue) {}
+
+	virtual void Draw() override;
+
```
Comment (TomGoodIdea) [inline source/NameDialog.h:38]:
```suggestion


```

File: source/NameDialog.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+protected:
+	virtual bool Click(int x, int y, int clicks) override;
+
```
Comment (TomGoodIdea) [inline source/NameDialog.h:41]:
```suggestion


```

File: source/NameDialog.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+private:
+	Point randomPos;
+};
+
```
Comment (TomGoodIdea) [inline source/NameDialog.h:45]:
```suggestion



```

**Review by TomGoodIdea — COMMENTED** [review]
style

File: source/NameDialog.h (hunk 5 above)
Comment (nickshanks) [reply source/NameDialog.h:27]: OK, but I copied this verbatim from the ShipyardPanel.h file!

File: source/NameDialog.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+*/
+
+#include "NameDialog.h"
+
+#include "Color.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+using namespace std;
```
Comment (TomGoodIdea) [inline source/NameDialog.cpp:23]:
```suggestion

using namespace std;
```
sorry, my fault

File: source/NameDialog.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+#include "Dialog.h"
+
+#include "GameData.h"
+#include "Phrase.h"
+#include "Point.h"
+
+#include <string>
+
+
```
Comment (TomGoodIdea) [inline source/NameDialog.h:27]:
```suggestion


```

File: source/NameDialog.cpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+void NameDialog::Draw()
+{
+	Dialog::Draw();
+
+	randomPos = cancelPos - Point(100., 0.);
+	SpriteShader::Draw(SpriteSet::Get("ui/dialog cancel"), randomPos);
+
+	const Font &font = FontSet::Get(14);
+	static const string RANDOM = "Suggest";
```
Comment (ziproot) [inline source/NameDialog.cpp:35]:
```suggestion
	static const string RANDOM = "Random";
```

This is the name for the same button when purchasing a new ship
image

Comment (nickshanks) [reply source/NameDialog.cpp:35]: The point of the PR is to change the name of this button + add it to the other rename dialog.

Comment (Amazinite) [reply source/NameDialog.cpp:36]: I'm in favor of adding this button to other locations, but I don't think it should be renamed.

File: source/NameDialog.h [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+// The name entry dialog should include a "Random" button to choose a random
+// name using the civilian ship name generator.
```
Comment (warp-core) [inline source/NameDialog.h:30]:
```suggestion
// A special version of Dialog for naming ships.
// Contains a text entry field and an additional button, "Random",
// which populates the text entry field with a randomly selected name
// from the "civilian" phrase.
```
I don't think the original comment is good enough if this class is going to be available globally.

File: source/NameDialog.h [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef NAME_DIALOG_H_
+#define NAME_DIALOG_H_
+
+#include "Dialog.h"
+
+#include "GameData.h"
+#include "Phrase.h"
```
Comment (warp-core) [inline source/NameDialog.h:22]: These aren't used in here.
```suggestion
```

File: source/NameDialog.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "NameDialog.h"
+
+#include "Color.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "SpriteSet.h"
```
Comment (warp-core) [inline source/NameDialog.cpp:21]:
```suggestion
#include "GameData.h"
#include "Phrase.h"
#include "SpriteSet.h"
```
They are used in here.

**Review by warp-core — COMMENTED** [review]
If this class is going in the global namespace, should it be renammed to something less ambiguous? Perhaps "ShipNameDialog"?

**Comment by TomGoodIdea** [comment]
@nickshanks can you address the review comments?

**Comment by nickshanks** [comment]
Hi, sorry, I've been busier again after a quiet period so haven't played ES in a few weeks. Did you mean answer the question about class name or merge the suggested changes?
The current button name isn't really suitable for reasons I described (it looks like it would close the dialog, not fill in the text field and leave the dialog open). If you don't like "Random" what else would work?

**Comment by warp-core** [comment]
I've applied the suggestions:
- change the label on the "random" button back to "Random" from "Suggest". I'm not convinced by the given argument for changing it, and I don't see anyone else in here supporting it, either. If you still feel it should be changed, feel free to open an issue/new PR, but the argument provided here so far is unlikely to result in such a change.
- update the comment on the class being moved from ShipyardPanel.cpp::<unnamed namespace> to the global namespace.
- Move some includes from the new header file to the new .cpp file.

I've also made some other changes:
- move the definition of the constructor from inside the class definition to after it, in the same file (because it is a template function) for style reasons.
- rename the "NameDialog" class to "ShipNameDialog" because I think the original name is too ambiguous when the class is available in the global namespace instead of just the anonymous namespace of ShipyardPanel.cpp.
- Change header guard style in the new header file from the old "#ifndef #define" method to "#pragma once".
- Update include paths for files that have been moved.

**Pull request merged**